nilearn/glm/first_level/first_level.py

```diff
@@ -897,58 +922,73 @@ def first_level_from_bids(dataset_path,
     # Throw warning if no bold.json is found
     if t_r is not None:
-        warn(f'RepetitionTime given in as {t_r}')
-        warn(f'slice_time_ref is {slice_time_ref} percent '
-             'of the repetition time')
+        _check_repetition_time(t_r)
+        warn("'RepetitionTime' given in model_init as {t_r}")
     else:
         filters = _make_bids_files_filter(
             task_label=task_label,
+            space_label=space_label,
             supported_filters=[*_bids_entities()["raw"], *_bids_entities()["derivatives"]],
-            extra_filter=img_filters
+            extra_filter=img_filters,
+            verbose=verbose
         )
-        img_specs = get_bids_files(derivatives_path,
-                                   modality_folder='func',
-                                   file_tag='bold',
-                                   file_type='json',
-                                   filters=filters)
-        # If we don't find the parameter information in the derivatives folder
-        # we try to search in the raw data folder
-        if not img_specs:
+        t_r = _infer_repetition_time_from_dataset(
+            bids_path=derivatives_path,
+            filters=filters,
+            verbose=verbose)
+        # If the parameter information is not found in the derivatives folder,
+        # a search is done in the raw data folder.         
+        if t_r is None:
             filters = _make_bids_files_filter(
                 task_label=task_label,
-                supported_filters=_bids_entities()["raw"],
-                extra_filter=img_filters
+                supported_filters=[*_bids_entities()["raw"]],
+                extra_filter=img_filters,
+                            verbose=verbose
             )
-            img_specs = get_bids_files(dataset_path,
-                                       modality_folder='func',
-                                       file_tag='bold',
-                                       file_type='json',
-                                       filters=filters)
-        if not img_specs:
-            warn('No bold.json found in the derivatives or dataset folder.'
-                 ' t_r can not be inferred '
-                 ' and will need to be set manually in the list of models'
-                 ' otherwise their fit will throw an exception.')
+            t_r = _infer_repetition_time_from_dataset(
+            bids_path=dataset_path,
+            filters=filters,
+            verbose=verbose)
+    if t_r is not None:        
+        _check_repetition_time(t_r)
+    else:
+        warn("'t_r' not provided and cannot be inferred from metadata. " 
+             "It will need to be set manually in the list of models, "
+             "otherwise their fit will throw an exception.")
+
+    if slice_time_ref is not None:
+        _check_slice_time_ref(slice_time_ref)
+        warn("'slice_time_ref' given in model_init as {slice_time_ref}")
```

Comment:
We should not warn here.

Reply:
Gonna try to make a small table about the different situations (input argument VS what metadata exist in which dataset) to clarify when to throw warnings or errors.

Reply:
This refactoring isn't very useful if `slice_time_ref` default is not changed to None. If we do this, we need to deprecate it properly as @bthirion pointed out. In any case, given the explanation in the docstrings, this warning becomes redundant with the one under it.

Reply:
will remove the warning

Reply:
@Remi-Gau Yes I agree that if we don't have to change the API, better not. But this implementation will only infer the slice_time_ref if the user passes None, while the old behavior tries to infer it if the default, which is `0.`, is kept. So a user that doesn't change their code that uses the default option will now get `0.` set as their slice_time_ref while they expect it to be inferred. Does that make sense? Or maybe I have the logic wrong. If what I'm saying follows, then we'd need to go through a deprecation cycle as well.

Reply:
If I am correct the old behavior was:
- try to infer `t_r` AND `slice_time_ref` **if and only if** `t_r = None` is passed (which was the default)
- if a `slice_time_ref` is found then it overrides ANY value passed as argument.

Making more tables to compare old and new behavior depending on datasets and arguments...